src/Symfony/Component/Security/Http/Firewall/ExceptionListener.php

```diff
@@ -100,6 +101,13 @@ private function handleAuthenticationException(GetResponseForExceptionEvent $eve
             $this->logger->info(sprintf('Authentication exception occurred; redirecting to authentication entry point (%s)', $exception->getMessage()));
         }
 
+        $session = $event->getRequest()->getSession();
+        $session->set(SecurityContextInterface::AUTHENTICATION_ERROR, $exception);
+
+        if ($exception instanceof UsernameNotFoundException) {
+            $session->set(SecurityContextInterface::LAST_USERNAME, $exception->getUsername());
+        }
```

Comment:
setting things in the session must not be done like this all the time:

- the request may not have a Session in it, which triggers a fatal error in your code
- if the firewall is configured as stateless, nothing should be set in the session even if there is a session